src/libraries/Common/src/System/Net/SocketProtocolSupportPal.cs

```diff
@@ -0,0 +1,63 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Sockets;
+
+namespace System.Net
+{
+    internal static partial class SocketProtocolSupportPal
+    {
+        private static bool s_ipv4;
+        private static bool s_ipv6;
+        private static bool s_uds;
+
+        private static readonly object s_initializedLock = new object();
+        private static volatile bool s_initialized;
+
+        public static bool OSSupportsIPv6
+        {
+            get
+            {
+                EnsureInitialized();
+                return s_ipv6;
+            }
+        }
+
+        public static bool OSSupportsIPv4
+        {
+            get
+            {
+                EnsureInitialized();
+                return s_ipv4;
+            }
+        }
+
+        public static bool OSSupportsUnixDomainSockets
+        {
+            get
+            {
+                EnsureInitialized();
+                return s_uds;
+            }
+        }
+
+        private static void EnsureInitialized()
+        {
+            if (!s_initialized)
+            {
+                lock (s_initializedLock)
```

Comment:
Why not just use a before-field-init static constructor for these?

Reply:
Good question.  I was just moving what was already there, but now that you point it out, I'll change it.